### PR TITLE
Use media kit as the default video playback for windows

### PR DESCRIPTION
**Pull request by bdlukaa**
I could experience gains of 30% to 40% in performance by using `media_kit` on Windows

## Features

- [x] Play from url
- [x] Play from file
- [x] Update volume
- [x] Update speed
- [x] Seeking

### Screenshots
image

---

There seems to be a minor flickering when reordering the views in the multiple view

### CODE REVIEW

File: packages/unity_video_player/unity_video_player_media_kit/lib/unity_video_player_media_kit.dart [hunk 1]
```diff
@@ -107,6 +101,17 @@ class __MKVideoState extends State<_MKVideo> {
 
 class UnityVideoPlayerMediaKit extends UnityVideoPlayer {
   Player mkPlayer = Player();
+  late Future<VideoController> mkVideoController;
```
Comment (alexmercerind) [inline packages/unity_video_player/unity_video_player_media_kit/lib/unity_video_player_media_kit.dart:104]: `VideoController` should be free'd in `dispose` (ideally before `Player` using `await`).

**Comment by bdlukaa** [comment]
I will keep this as a draft because it's causing crashes to the app. `media_kit` is not stable enough to give it a try for now

¹ For some reason, it seems it's blocking the UI thread, so app behaves extremely slow. Sometimes, it's unusable
² When navigating to another screens, the `VideoController` object is kept alive (so that is a gapless playback when coming back to the grid). This seems to be causing crashes since the Video object is not into view

cc @alexmercerind

**Comment by alexmercerind** [comment]
> For some reason, it seems it's blocking the UI thread, so app behaves extremely slow. 

The video rendering does occur on render thread. A "hit" due to multiple videos will definitely be there. If there are no videos in the view, there should be no frame drops/delays etc.

<hr>

**Crash:**

There shouldn't be _any crashes_ on Windows. I & other users have confirmed the recent fix which addressed last few race(s). There are a number of cases/situations in the test application where we haven't been able to see any crash since latest fixes.

As for _any crash_, let's see what further feedback I get. Personally, I'm unable to.

**Performance:**


There isn't any scope of improvement in terms of performance AFAIK. In comparison to package:dart_vlc, the performance offered by package:media_kit is already very-high. Everything is being rendered off the GPU.

**Comment by curtishall** [comment]
@alexmercerind I know you are busy, when you have time can you review this PR?  It's blocking #31

**Review by alexmercerind — APPROVED** [review]
LGTM (+ recommendations) !

You can try use the current package:media_kit branch that I'm working on.

https://github.com/alexmercerind/media_kit/tree/perf/win32-copy-separate-id3d11texture2d

The level of performance should be further improved as compared to `main` branch. There are few more performance improvements which I'd like to make, however I won't be able to get this branch/PR over the line before this month (though it's usable in current state).

<hr></hr>

As far as https://github.com/alexmercerind/media_kit/issues/28 is concerned, I don't think it will ever be. You have to create your own solution here. Primarily because libmpv does not offer such functionality out-of-the-box & implementing one from scratch will be:
- Redundant.
- Requiring more maintenance cost.
- Of different level of abstraction that package:media_kit aims to offer (if custom implementation is made).

100% of implementation of package:media_kit is in pure Dart, only video rendering is platforms specific C/C++. You're free to play around.

<hr></hr>


> There seems to be a minor flickering when reordering the views in the multiple view.

I see the last commit with `ValueKey`, that should improve IG. I'm not sure however.


<hr></hr>

One final recommendation I'd like to give is that you can try to specify `width` & `height` parameters in the `VideoController.create` method. It will create a less sized D3D Texture / EGLSurface (and thus squeeze the performance).

Especially in grid layouts, where video won't be visible in large sizes, a fixed `width` & `height` texture will work & take less resources. However, when viewing same camera on larger screen, a different `VideoController` without fixed `width` & `height` can be used (& disposed upon exit).

**Comment by bdlukaa** [comment]
> One final recommendation I'd like to give is that you can try to specify width & height parameters in the VideoController.create method. It will create a less sized D3D Texture / EGLSurface (and thus squeeze the performance).

Indeed. By doing that, I got a ~15% performance improvement

**Pull request merged**